File: modules/image-palette/ios/PaletteCache.swift
import Foundation

struct PaletteKey: Hashable {
  let uri: String
  let gridWidth: Int
  let gridHeight: Int
  let edgesOnly: Bool
  let bitsPerChannel: Int
  let downsample: Bool
  let downsampleTargetSize: Int

  init(uri: String, config: Config) {
    self.uri = uri
    self.gridWidth = config.gridWidth
    self.gridHeight = config.gridHeight
    self.edgesOnly = config.edgesOnly
    self.bitsPerChannel = config.bitsPerChannel
    self.downsample = config.downsample
    self.downsampleTargetSize = config.downsampleTargetSize
  }
}

final class PaletteCache {
  private var store: [PaletteKey: [InternalSwatch]] = [:]

  func get(_ key: PaletteKey) -> [InternalSwatch]? {
    return store[key]
  }

  func set(_ key: PaletteKey, _ value: [InternalSwatch]) {
    store[key] = value
  }

  func clear() {
    store.removeAll()
  }
}

File: modules/image-palette/android/src/main/java/expo/modules/imagepalette/PaletteCache.kt
package expo.modules.imagepalette

data class PaletteKey(
  val uri: String,
  val gridWidth: Int,
  val gridHeight: Int,
  val edgesOnly: Boolean,
  val bitsPerChannel: Int,
  val downsample: Boolean,
  val downsampleTargetSize: Int
) {
  companion object {
    fun fromConfig(uri: String, config: Config) = PaletteKey(
      uri = uri,
      gridWidth = config.gridWidth,
      gridHeight = config.gridHeight,
      edgesOnly = config.edgesOnly,
      bitsPerChannel = config.bitsPerChannel,
      downsample = config.downsample,
      downsampleTargetSize = config.downsampleTargetSize
    )
  }
}

class PaletteCache {
  private val store = mutableMapOf<PaletteKey, List<InternalSwatch>>()

  fun get(key: PaletteKey): List<InternalSwatch>? = store[key]

  fun set(key: PaletteKey, value: List<InternalSwatch>) {
    store[key] = value
  }

  fun clear() {
    store.clear()
  }
}
